src/NLog/Internal/TargetWithFilterChain.cs

```diff
@@ -50,6 +51,115 @@ internal class TargetWithFilterChain
         /// </summary>
         private StackTraceUsage? _stackTraceUsage;
 
+        struct CallSiteKey : IEquatable<CallSiteKey>
+        {
+            public CallSiteKey(string methodName, string fileSourceName, int fileSourceLineNumber)
+            {
+                MethodName = methodName ?? string.Empty;
+                FileSourceName = fileSourceName ?? string.Empty;
+                FileSourceLineNumber = fileSourceLineNumber;
+            }
+
+            public readonly string MethodName;
+            public readonly string FileSourceName;
+            public readonly int FileSourceLineNumber;
+
+            /// <summary>
+            /// Serves as a hash function for a particular type.
+            /// </summary>
+            /// <returns>
+            /// A hash code for the current <see cref="T:System.Object"/>.
+            /// </returns>
+            public override int GetHashCode()
+            {
+                return MethodName.GetHashCode() ^ FileSourceName.GetHashCode() ^ FileSourceLineNumber;
+            }
+
+            /// <summary>
+            /// Determines if two objects are equal in value.
+            /// </summary>
+            /// <param name="obj">Other object to compare to.</param>
+            /// <returns>True if objects are equal, false otherwise.</returns>
+            public override bool Equals(object obj)
+            {
+                return obj is CallSiteKey key && Equals(key);
+            }
+
+            /// <summary>
+            /// Determines if two objects of the same type are equal in value.
+            /// </summary>
+            /// <param name="other">Other object to compare to.</param>
+            /// <returns>True if objects are equal, false otherwise.</returns>
+            public bool Equals(CallSiteKey other)
+            {
+                return FileSourceLineNumber == other.FileSourceLineNumber
```

Comment:
I think it's missing the null comparisons, See https://docs.microsoft.com/en-us/previous-versions/ms173147(v=vs.90)

PS: Resharper could generate equals and hashcode

Reply:
CallSiteKey is a struct, so no need for null-check

Reply:
ah missed that, thanks